core/src/main/java/org/elasticsearch/ingest/processor/SetProcessor.java

```diff
@@ -55,7 +65,9 @@ public ValueSource getValue() {
 
     @Override
     public void execute(IngestDocument document) {
-        document.setFieldValue(field, value);
+        if (overrideEnabled || !document.hasField(field) || document.getFieldValue(field, Object.class) == null) {
```

Comment:
Can you change `!document.hasField(field)` into `document.hasField(field) == false`. This is kind of the standard now in the code base, because it is easy to overlook the `!`.